### PR TITLE
Add support for tracking metadata

### PR DESCRIPTION
**Pull request by palemieux**

### CODE REVIEW

File: src/main/python/camdkit/framework.py [hunk 1]
```diff
@@ -16,6 +16,7 @@
 class Sampling(Enum):
   STATIC = "Static"
   REGULAR = "Regular"
+  DYNAMIC = "Dynamic"
```
Comment (palemieux) [inline src/main/python/camdkit/framework.py:19]: What is the difference between "Dynamic" and "Regular"?

**Comment by palemieux** [comment]
@jamesmosys Why create a separate framework for tracking metadata? Have you considered instead adding a label to each metadata item, indicating the context(s) in which it could be used?

File: src/main/python/camdkit/framework.py (hunk 1 above)
Comment (jamesmosys) [reply src/main/python/camdkit/framework.py:19]: Regular seemed to be a case that was used for arrays so I created a new case so as not to break any of that...

Comment (palemieux) [reply src/main/python/camdkit/framework.py:19]: What is dynamic then?

**Comment by palemieux** [comment]
> Should I move some of the more generic classes to the camdkit/framework.py? I could move TrackingParameter to the trackerkit/model.py and then remove trackerkit/framework.py altogether...

I do not see a reason to have two frameworks, assuming the data models are similar. We might want to add a label tag to each parameter to identify to which domain of application it applies.

File: src/main/python/camdkit/framework.py (hunk 1 above)
Comment (jamesmosys) [reply src/main/python/camdkit/framework.py:19]: Ah I did not understand regularly sampled. I took regular to mean 'normal' - oops! I added the DYNAMIC case so as not to touch the existing code too much. (I added recursion on ParameterContainer::to_json e.g.). Perhaps the framework needs to be refactored to handle objects and arrays in a more generic way? I'm happy to have a go at this, but is it a separate (later) PR?

**Comment by jamesmosys** [comment]
> > Should I move some of the more generic classes to the camdkit/framework.py? I could move TrackingParameter to the trackerkit/model.py and then remove trackerkit/framework.py altogether...
> 
> I do not see a reason to have two frameworks, assuming the data models are similar. We might want to add a label tag to each parameter to identify to which domain of application it applies.

Ok I will migrate the framework classes I added into the main framework

File: src/main/python/camdkit/framework.py (hunk 1 above)
Comment (palemieux) [reply src/main/python/camdkit/framework.py:19]: > Perhaps the framework needs to be refactored to handle objects and arrays in a more generic way?

A [parameter](https://github.com/SMPTE/ris-osvp-metadata-camdkit/blob/0e56559d7d53e0c9a585addc676c64afaca1e065/src/main/python/camdkit/framework.py#L27) can be anything, including an array. What do you mean by "more generic"?

>  I'm happy to have a go at this, but is it a separate (later) PR?

Yes.

Comment (jamesmosys) [reply src/main/python/camdkit/framework.py:19]: > > Perhaps the framework needs to be refactored to handle objects and arrays in a more generic way?
> 
> A [parameter](https://github.com/SMPTE/ris-osvp-metadata-camdkit/blob/0e56559d7d53e0c9a585addc676c64afaca1e065/src/main/python/camdkit/framework.py#L27) can be anything, including an array. What do you mean by "more generic"?

Not currently - if I just set my parameters to regular then lines like this won't work:
```
        elif desc.sampling is Sampling.REGULAR:
          self._values[prop] = tuple(map(desc.from_json, json_value))
```

Comment (palemieux) [reply src/main/python/camdkit/framework.py:19]: Each parameter in the model must be defined in https://github.com/SMPTE/ris-osvp-metadata-camdkit/blob/main/src/main/python/camdkit/model.py

Comment (palemieux) [reply src/main/python/camdkit/framework.py:19]: See usage at https://github.com/SMPTE/ris-osvp-metadata-camdkit/blob/main/src/test/python/test_model.py

Comment (jamesmosys) [reply src/main/python/camdkit/framework.py:19]: Yes I have created my parameters in trackerkit/model.py - I will look at the tests to see what I was doing wrong. Thanks.

Comment (jamesmosys) [reply src/main/python/camdkit/framework.py:19]: Ok I get it now. A Clip has tuples of 'regular' data. Whereas my 'Frame' has a single data point, but it is not 'static' as such. So I can now answer your original question. My concept of 'Dynamic' data is one that is changing every one of my 'Frames' I will comment the enum accordingly. We may want to rethink this in the future, but are you happy to proceed with the 'Dynamic' option for now?

Comment (jamesmosys) [reply src/main/python/camdkit/framework.py]: Just to close off this thread - after much discussion on Slack I will be reworking the architecture of this to integrate better with the existing model, rather than work alongside it.

File: src/main/python/camdkit/framework.py [hunk 2]
```diff
@@ -334,24 +448,34 @@ def from_json(self, json_dict: dict):
...
   @classmethod
-  def make_json_schema(cls) -> dict:
-    schema = {
-      "$schema": "https://json-schema.org/draft/2020-12/schema",
-      "type": "object",
-      "properties": {}
-    }
+  def make_json_schema(cls, is_root=False) -> dict:
+    schema = {}
+    if is_root:
```
Comment (palemieux) [inline src/main/python/camdkit/framework.py:453]: Why is this needed?

The framework current assumes a flat model where a `ParameterContainer` contains multiple `Parameter` but cannot contain other `ParameterContainer`.

**Comment by palemieux** [comment]
@jamesmosys The model currently supported by camdkit is one where a `Clip`, which is a `ParameterContainer`, can contain any number of `Parameter`, each optional and each either static or regularly sampled. A `Parameter` cannot contain other `Parameter`.

What about adding the tracking parameters to the `Clip` like other camera parameters? Any downsides?

File: src/main/python/camdkit/framework.py (hunk 2 above)
Comment (jamesmosys) [reply src/main/python/camdkit/framework.py:453]: Ah yes this is a leftover from when I implemented recursive nesting here - will remove.

File: src/main/python/camdkit/framework.py [hunk 3]
```diff
@@ -257,6 +277,98 @@ def make_json_schema() -> dict:
       "maximum": 2147483647
     }
 
+class TrackingParameter(Parameter):
```
[the remark this answers is not in the document]
Comment (jamesmosys) [reply src/main/python/camdkit/framework.py:280]: Agreed this is no longer required since the whole framework has been consolidated (it made more sense when there was a trackerkit/framework)

File: src/main/python/camdkit/trackerkit/model.py
[the remark this answers is not in the document]
Comment (jamesmosys) [reply src/main/python/camdkit/trackerkit/model.py:1]: I have to do some swizzling in trackerkit/model to present the json appropriately for live data. I convert the representation as follows:
```
   # from:
   obj: {
      param1: [a, b, c]
   }
   # to:
   [{ obj: { 
      param1: a
   }, obj: {
      param1: b
   }, obj: {
      param1: c
   }]
```
This is not exactly pretty, but it does mean that the underlying core data storage model is CamDKit, and it is only in the the presentation of the data that it gets swizzled.
WRT moving the fields into the model - I think that would be fine - I'll just inherit TrackingClip from Clip instead of ParameterContainer. Only downside I can see is it will make model.py quite bloated as we have a lot of fields. But I'll give that a go today.

File: src/main/python/camdkit/framework.py [hunk 4]
```diff
@@ -310,10 +422,12 @@ def to_json(self) -> dict:
     obj = {}
     for k, desc in self._params.items():
       value = self._values[k]
+      # TODO JU how to handle optional parameters?
```
[the remark this answers is not in the document]
Comment (jamesmosys) [reply src/main/python/camdkit/framework.py]: NVM this was my bug. Fixed

**Review by palemieux — COMMENTED** [review]
Is src/main/python/camdkit/trackerkit/__init__.py still needed?

File: src/main/python/camdkit/framework.py [hunk 5]
```diff
@@ -233,30 +312,338 @@ def make_json_schema() -> dict:
...
+    """The parameter shall be one of the allowed values."""
+    return str(value) in self.allowedValues
+
+  def make_json_schema(self) -> dict:
+    return {
+      "type": "string",
+      "enum": self.allowedValues
+    }
+  
+class TimingModeParameter(EnumParameter):
```
Comment (palemieux) [inline src/main/python/camdkit/framework.py:402]: No need to define this class when there is a single derived class. Also, the sampling should be specified in the model so that the same type can be used for both dynamic and static parameters.

File: src/main/python/camdkit/framework.py [hunk 6]
```diff
@@ -233,30 +312,338 @@ def make_json_schema() -> dict:
...
+    return float(value)
+  
+  @staticmethod
+  def make_json_schema() -> dict:
+    return {
+      "type": "number",
+      "minimum": 0.0,
+    }
+  
+class EnumParameter(StringParameter):
```
Comment (palemieux) [inline src/main/python/camdkit/framework.py:381]: Shouldn't this be derived from Python enums? In any case, why is `enum_class` necessary since the name of the derived class can be accessed?

File: Pipfile [hunk 7]
```diff
@@ -7,5 +7,8 @@ verify_ssl = true
 pylint = "==2.6.0"
 coverage = "*"
 
+[packages]
+jsonschema = "*"
```
Comment (palemieux) [inline Pipfile:11]: This should be a dev package.

File: src/main/python/camdkit/framework.py [hunk 8]
```diff
@@ -233,30 +312,338 @@ def make_json_schema() -> dict:
...
+        },
+        "format": {
+          "type": "string",
+          "enum": ["24", "24D", "25", "30", "30D"]
+        }
+      }
     }
 
+class TransformsParameter(Parameter):
+  sampling = Sampling.REGULAR
```
Comment (palemieux) [inline src/main/python/camdkit/framework.py:528]: The sampling is set in the model.

File: src/main/python/camdkit/framework.py [hunk 9]
```diff
@@ -233,30 +312,338 @@ def make_json_schema() -> dict:
...
+          "minimum": 0,
+          "maximum": UINT_MAX
+        }
+      },
+      "required": ["seconds", "nanoseconds"]
+    }
+
+
+class TimecodeParameter(Parameter):
+  sampling = Sampling.REGULAR
```
Comment (palemieux) [inline src/main/python/camdkit/framework.py:460]: `sampling` is set in `model`. The idea is that any parameter can be used either statically or dynamically.

File: src/main/python/camdkit/framework.py [hunk 10]
```diff
@@ -364,6 +773,32 @@ def make_documentation(cls) -> dict:
         "description" : desc.__doc__,
         "constraints" : desc.validate.__doc__,
         "sampling" : str(desc.sampling.value),
-        "units": desc.units
+        "units": desc.units if hasattr(desc, "units") else "None"
```
Comment (palemieux) [inline src/main/python/camdkit/framework.py:776]: All items must have `units` specified in `model.py`, even `None`.

File: src/main/python/camdkit/framework.py [hunk 11]
```diff
@@ -364,6 +773,32 @@ def make_documentation(cls) -> dict:
...
       })
     return doc
+  
+  """
```
Comment (palemieux) [inline src/main/python/camdkit/framework.py:780]: `ParameterContainer` is not a temporal container, but a container of metadata parameters. What is this iteration intended to do?

File: src/main/python/camdkit/model.py [hunk 12]
```diff
@@ -194,6 +192,60 @@ def make_json_schema() -> dict:
...
+  Euler angles are human readable and unlike quarternions, provide the ability for cycles (with angles >360 or <0 degrees).
+  Where a tracking system is providing the pose of a virtual camera, gimbal lock does not present the physical challenges of a robotic system.
+  Conversion to and from quarternions is trivial with an acceptable loss of precision
+  """
+  canonical_name = "transforms"
+  units = "metres / degrees"
+
+class TimingMode(TimingModeParameter):
+  """
+  'external' timing mode describes the case where the transport packet has inherent timing, so no explicit timing data is required in the data).
```
Comment (palemieux) [inline src/main/python/camdkit/model.py:213]: This is not needed: either a PTP timestamp is present or it is not.

File: src/main/python/camdkit/model.py [hunk 13]
```diff
@@ -117,6 +114,7 @@ class FStop(StrictlyPositiveIntegerParameter):
   canonical_name = "fStop"
   sampling = Sampling.REGULAR
   units = "0.001 unit"
+  section = "lens"
```
Comment (palemieux) [inline src/main/python/camdkit/model.py:117]: Why is the only parameter moved to `lens`?

File: src/main/python/camdkit/model.py (hunk 12 above)
Comment (jamesmosys) [reply src/main/python/camdkit/model.py:213]: The idea of this was that you can still use PTP timing even if the transport you are in doesn't have it inherently. Can you think of a better name? Intrinsic / extrinsic?

File: src/main/python/camdkit/model.py (hunk 13 above)
Comment (jamesmosys) [reply src/main/python/camdkit/model.py:117]: More will come as I implement the rest of the model

File: src/main/python/camdkit/framework.py (hunk 10 above)
Comment (jamesmosys) [reply src/main/python/camdkit/framework.py:776]: This is not enforced currently (sampling and canonical name are) - shall I update ParameterContainer.__init_subclass__ to enforce units?

File: src/main/python/camdkit/framework.py (hunk 6 above)
Comment (jamesmosys) [reply src/main/python/camdkit/framework.py:381]: It is represented as a string in json schema with an enum [...] property. (The enum_class does derive from the Python Enum.) I will determine the enum_class from the class name as you suggest.

File: src/main/python/camdkit/framework.py (hunk 5 above)
Comment (jamesmosys) [reply src/main/python/camdkit/framework.py:402]: I've moved them to the model if there is only a single derived class

File: src/main/python/camdkit/framework.py (hunk 11 above)
Comment (jamesmosys) [reply src/main/python/camdkit/framework.py:540]: This is how I achieve the rearranging of the data for display as per the comment above:
```
 # from:
   obj: {
      param1: [a, b, c]
   }
   # to:
   [{ obj: { 
      param1: a
   }, obj: {
      param1: b
   }, obj: {
      param1: c
   }]
```

File: src/main/python/camdkit/model.py (hunk 12 above)
Comment (palemieux) [reply src/main/python/camdkit/model.py]: Why is knowing the source of the PTP timestamp important?

File: src/main/python/camdkit/framework.py [hunk 14]
```diff
@@ -7,22 +7,188 @@
...
+@dataclasses.dataclass
+class Orientations:
+  "Horizontal and vertical measurements"
+  horizontal: numbers.Real
+  vertical: numbers.Real
+
+@dataclasses.dataclass
+class Vector3:
+  "3 doubles x,y,z to encode - for example - a location or a translation."
+  x: typing.Optional[numbers.Real] = 0.0
```
Comment (palemieux) [inline src/main/python/camdkit/framework.py:37]: In what circumstance(s) would a coordinate of a vector be optional?

File: src/main/python/camdkit/framework.py [hunk 15]
```diff
@@ -7,22 +7,188 @@
...
+  y: typing.Optional[numbers.Real] = 0.0
+  z: typing.Optional[numbers.Real] = 0.0
+
+@dataclasses.dataclass
+class Rotator3:
+  "3 doubles pan, tilt, roll to encode - for example - a camera rotation."
+  pan: typing.Optional[numbers.Real] = 0.0
```
Comment (palemieux) [inline src/main/python/camdkit/framework.py:44]: I recommend against picking default values and instead force the application to explicitly specify values.

File: src/main/python/camdkit/framework.py [hunk 16]
```diff
@@ -7,22 +7,188 @@
...
+  
+  @staticmethod
+  def to_json(value: typing.Any) -> typing.Any:
+    return dataclasses.asdict(value)
+
+class TimingModeEnum(BaseEnum):
+  INTERNAL = "internal"
+  EXTERNAL = "external"
+
+class TimecodeFormat(BaseEnum):
```
Comment (palemieux) [inline src/main/python/camdkit/framework.py:145]: The enumeration is missing frame rates.

In order to fully parameterize timecode, you need two parameters: frame rate and whether the timecode is drop frame or non drop frame.

File: src/main/python/camdkit/framework.py [hunk 17]
```diff
@@ -7,22 +7,188 @@
...
+  def to_int(cls, value):
+    if value == cls.TC_24 or value == cls.TC_24D: return 24
+    if value == cls.TC_25: return 25
+    if value == cls.TC_30 or value == cls.TC_30D: return 30
+    raise ValueError
+  
+  @classmethod
+  def to_float(cls, value):
+    if value == cls.TC_24: return 24.0
+    if value == cls.TC_24D: return 23.976
```
Comment (palemieux) [inline src/main/python/camdkit/framework.py:162]: This should return a rational number and not an inaccurate decimal number.

File: src/main/python/camdkit/framework.py [hunk 18]
```diff
@@ -7,22 +7,188 @@
...
+  hours: int
+  minutes: int
+  seconds: int
+  frames: int
+  format: TimecodeFormat
+
+@dataclasses.dataclass
+class Synchronization:
+  "Data structure for synchronisation data"
+  frequency: float
```
Comment (palemieux) [inline src/main/python/camdkit/framework.py:184]: What are the units of frequency? Is this the nominal frequency of the time source? How is it used?

File: src/main/python/camdkit/framework.py [hunk 19]
```diff
@@ -7,22 +7,188 @@
...
+  locked: bool
```
Comment (palemieux) [inline src/main/python/camdkit/framework.py:185]: Locked to what?

File: src/main/python/camdkit/framework.py [hunk 20]
```diff
@@ -7,22 +7,188 @@
...
+  source: SynchronizationSourceEnum
+  ptp_master: typing.Optional[str] = None
```
Comment (palemieux) [inline src/main/python/camdkit/framework.py:187]: Isn't this mutually exclusive with `frequency` and only relevant if `source` is PTP? If so, I recommend splitting this into different structures, one for each kind of time source.

File: src/main/python/camdkit/framework.py [hunk 21]
```diff
@@ -7,22 +7,188 @@
...
+  ptp_offset: typing.Optional[float] = None
+  ptp_domain: typing.Optional[int] = None
+  offsets: typing.Optional[SynchronizationOffsets] = None
+  enabled: typing.Optional[bool] = None
```
Comment (palemieux) [inline src/main/python/camdkit/framework.py:191]: What does `enabled` mean?

File: src/main/python/camdkit/framework.py (hunk 11 above)
Comment (palemieux) [reply src/main/python/camdkit/framework.py:743]: A `ParameterContainer` can only be iterated on if it consists only of `REGULAR` parameters and only if all such parameters are sampled at the same rate. I recommend removing this method and leaving to the application.

File: src/main/python/camdkit/model.py [hunk 22]
```diff
@@ -36,6 +35,7 @@ class CaptureFPS(StrictlyPositiveRationalParameter):
   canonical_name = "captureRate"
   sampling = Sampling.STATIC
   units = "hertz"
+  section = "camera"
```
Comment (palemieux) [inline src/main/python/camdkit/model.py:38]: Doesn't a tracking capture also happen over a duration? If `duration` is not available for tracking metadata, how is the sampling frequency of each metadata parameter determined?

If tracking metadata is not necessarily sampled regularly, and instead each sample contains a timestamp, then we should probably use something other than `ParameterContainer`.

File: src/main/python/camdkit/model.py [hunk 23]
```diff
@@ -52,55 +53,669 @@ class LensSerialNumber(StringParameter):
   canonical_name = "lensSerialNumber"
   sampling = Sampling.STATIC
   units = None
+  section = "lens"
```
Comment (palemieux) [inline src/main/python/camdkit/model.py:56]: Including _lens_ in the name and having a section called _lens_ is redundant.

We should decide to group parameters by name, explicit section, or tags.

File: src/main/python/camdkit/model.py [hunk 24]
```diff
@@ -52,55 +53,669 @@ class LensSerialNumber(StringParameter):
...
+
+  @staticmethod
+  def make_json_schema() -> dict:
+    return {
+      "type": "integer",
+      "minimum": 1,
+      "maximum": 360000
+    }
+  
+class PacketId(UUIDURNParameter):
```
Comment (palemieux) [inline src/main/python/camdkit/model.py:197]: This is probably the id of the sample, right? I think the distinction between sample and packet is important since the transport might fragment a sample across multiple packets, or aggregate multiple samples in one packet.

File: src/main/python/camdkit/model.py [hunk 25]
```diff
@@ -117,23 +733,23 @@ class FStop(StrictlyPositiveIntegerParameter):
   canonical_name = "fStop"
   sampling = Sampling.REGULAR
   units = "0.001 unit"
+  section = "lens"
 
-class FocalLength(StrictlyPositiveIntegerParameter):
-  """Nominal focal length of the lens. The number printed on the side of a prime
-  lens, e.g. 50 mm, and undefined in the case of a zoom lens."""
+class FocalLength(NonNegativeRealParameter):
+  """Focal length of the lens."""
```
Comment (palemieux) [inline src/main/python/camdkit/model.py:739]: The long description text should go back in since it was carefully crafted.

Comment (jamesmosys) [reply src/main/python/camdkit/model.py:757]: Added a separate field NominalFocalLength for this

File: src/main/python/camdkit/model.py [hunk 26]
```diff
@@ -44,63 +60,768 @@ class ISO(StrictlyPositiveIntegerParameter):
...
+          "type": "integer",
+          "minimum": 0
+        }
+      },
+      "anyOf": [ {"required": ["focus"]}, {"required": ["iris"]}, {"required": ["zoom"]}
+      ]
+    }
+  
+  
+class TimingMode(EnumParameter):
```
Comment (JGoldstone) [inline src/main/python/camdkit/model.py:700]: Isn't this a semantically equivalent copy of what's at 473:482 (with the only difference being the wording of the doctoring)?

File: src/main/python/camdkit/model.py (hunk 23 above)
Comment (jamesmosys) [reply src/main/python/camdkit/model.py:72]: This is complete in latest commits

File: OpenTrackIO_WIP.json [hunk 27]
```diff
@@ -0,0 +1,265 @@
+# Example JSON packets for OpenTrackIO - camera tracking data for OSVP
+
+# Notes:
```
[the remark this answers is not in the document]
Comment (jamesmosys) [reply OpenTrackIO_WIP.json:3]: This is just a temp file for reviewing. It will be removed

File: src/main/python/camdkit/framework.py (hunk 14 above)
Comment (jamesmosys) [reply src/main/python/camdkit/framework.py:37]: If a mechanical encoder only report x and y, but has no z. (This is different from reporting z=0)

File: src/main/python/camdkit/framework.py (hunk 11 above)
Comment (jamesmosys) [reply src/main/python/camdkit/framework.py:771]: I don't currently have an alternative way of doing this. Can we discuss?

File: src/main/python/camdkit/model.py (hunk 26 above)
Comment (jamesmosys) [reply src/main/python/camdkit/model.py:718]: Please can you clarify - I think the line numbers have changed since this comment. Thanks!

**Review by jamesmosys — APPROVED** [review]
Ready to merge - outstanding issues moved to gitlab issues
